drivers/adc/adc_b91.c

```diff
@@ -0,0 +1,410 @@
+/*
+ * Copyright (c) 2022 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+#include "adc.h"
+#include <zephyr/dt-bindings/adc/b91-adc.h>
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_b91);
+
+#define DT_DRV_COMPAT telink_b91_adc
+
+/* ADC B91 defines */
+#define SIGN_BIT_POSITION          (13)
+#define AREG_ADC_DATA_STATUS       (0xf6)
+#define ADC_DATA_READY             (0x01)
+
+/* B91 ADC driver data */
+struct driver_data {
+	struct adc_context ctx;
+	int16_t *buffer;
+	int16_t *repeat_buffer;
+	uint8_t differential;
+	uint8_t resolution_offset;
+	struct k_sem acq_sem;
+	struct k_thread thread;
+
+	K_THREAD_STACK_MEMBER(stack, CONFIG_ADC_B91_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+static struct driver_data m_data = {
+	ADC_CONTEXT_INIT_TIMER(m_data, ctx),
+	ADC_CONTEXT_INIT_LOCK(m_data, ctx),
+	ADC_CONTEXT_INIT_SYNC(m_data, ctx),
+};
+
+/* Validate ADC data buffer size */
+static int adc_b91_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	size_t needed = sizeof(int16_t);
+
+	if (sequence->options) {
+		needed *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+/* Validate ADC read API input parameters */
+static int adc_b91_validate_sequence(const struct adc_sequence *sequence)
+{
+	int status;
+
+	if (sequence->channels != BIT(0)) {
+		LOG_ERR("Only channel 0 is supported.");
+		return -ENOTSUP;
+	}
+
+	if (sequence->oversampling) {
+		LOG_ERR("Oversampling is not supported.");
+		return -ENOTSUP;
+	}
+
+	status = adc_b91_validate_buffer_size(sequence);
+	if (status) {
+		LOG_ERR("Buffer size too small.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/* Convert dts pin to B91 SDK pin */
+static adc_input_pin_def_e adc_b91_get_pin(uint8_t dt_pin)
+{
+	adc_input_pin_def_e adc_pin;
+
+	switch (dt_pin) {
+	case DT_ADC_GPIO_PB0:
+		adc_pin = ADC_GPIO_PB0;
+		break;
+	case DT_ADC_GPIO_PB1:
+		adc_pin = ADC_GPIO_PB1;
+		break;
+	case DT_ADC_GPIO_PB2:
+		adc_pin = ADC_GPIO_PB2;
+		break;
+	case DT_ADC_GPIO_PB3:
+		adc_pin = ADC_GPIO_PB3;
+		break;
+	case DT_ADC_GPIO_PB4:
+		adc_pin = ADC_GPIO_PB4;
+		break;
+	case DT_ADC_GPIO_PB5:
+		adc_pin = ADC_GPIO_PB5;
+		break;
+	case DT_ADC_GPIO_PB6:
+		adc_pin = ADC_GPIO_PB6;
+		break;
+	case DT_ADC_GPIO_PB7:
+		adc_pin = ADC_GPIO_PB7;
+		break;
+	case DT_ADC_GPIO_PD0:
+		adc_pin = ADC_GPIO_PD0;
+		break;
+	case DT_ADC_GPIO_PD1:
+		adc_pin = ADC_GPIO_PD1;
+		break;
+	case DT_ADC_VBAT:
+		adc_pin = ADC_VBAT;
+		break;
```

Comment:
I'm wondering - couldn't this be somehow shortened? I.e. by using the `FOR_EACH` macro, because almost all of the cases in this switch statement seem to write the matched value to the `adc_pin` variable.

Reply:
Done.